capability.go

```diff
@@ -127,6 +130,55 @@ type clientState struct {
 	}
 }
 
+// clientCursor is an indirection pointing to a link in the resolution
+// chain of clientHooks. Places that need to do path shortening should
+// store one of these, rather than storing clientHook directly.
+type clientCursor struct {
+	hook *rc.Ref[clientHook] // nil if resolved to nil or released
+}
+
+func newClientCursor(hook clientHook) *rc.Ref[mutex.Mutex[clientCursor]] {
+	return rc.NewRefInPlace(func(c *mutex.Mutex[clientCursor]) func() {
+		c.With(func(c *clientCursor) {
+			c.hook = rc.NewRefInPlace(func(h *clientHook) func() {
+				*h = hook
+				return h.Release
+			})
+		})
+		return func() {
+			c.With(func(c *clientCursor) {
+				c.hook.Release()
+			})
+		}
+	})
+}
+
+// compress advances the hook referred to by this cursor as far
+// as possible without blocking on a resolution.
+func (c *clientCursor) compress() {
+	for {
+		if c.hook == nil {
+			return
+		}
+		res, ok := c.hook.Value().resolution.Get()
+		if !ok {
+			return
+		}
+		l := res.Lock()
+		if !l.Value().isResolved() {
+			l.Unlock()
+			return
+		}
+		r := l.Value().resolvedHook
+		if r != nil {
+			r = r.AddRef()
```

Comment:
Why do we need the `AddRef` here?

Reply:
The call to c.hook.Release() below will drop its reference to resolveHook, so we need to get our own reference here to make sure that doesn't spuriously release the hook.